c_api/src/android.rs

```diff
@@ -0,0 +1,477 @@
+            },
+            Err(_) => return 0,
+        }
+    };
+}
+
+/// Helper macro to handle optional Java string conversion for jstring return types
+macro_rules! java_string_to_cstring_opt_jstring {
+    ($env:expr, $jstring:expr) => {
+        if $jstring.is_null() {
+            None
+        } else {
+            match $env.get_string(&$jstring.into()) {
+                Ok(java_str) => match CString::new(java_str.to_str().unwrap_or("")) {
+                    Ok(c_str) => Some(c_str),
+                    Err(_) => return ptr::null_mut(),
+                },
+                Err(_) => return ptr::null_mut(),
+            }
+        }
+    };
+}
+
+/// Helper macro to handle optional Java string conversion for jlong return types
+macro_rules! java_string_to_cstring_opt_jlong {
+    ($env:expr, $jstring:expr) => {
+        if $jstring.is_null() {
+            None
+        } else {
+            match $env.get_string(&$jstring.into()) {
+                Ok(java_str) => match CString::new(java_str.to_str().unwrap_or("")) {
+                    Ok(c_str) => Some(c_str),
+                    Err(_) => return 0,
+                },
+                Err(_) => return 0,
+            }
+        }
+    };
+}
+
+/// Helper function to convert C string to Java string
+fn c_string_to_java_string(env: &mut JNIEnv, c_str: *mut c_char) -> jstring {
+    if c_str.is_null() {
+        return ptr::null_mut();
+    }
+    
+    let rust_str = unsafe {
+        match CStr::from_ptr(c_str).to_str() {
+            Ok(s) => s,
+            Err(_) => return ptr::null_mut(),
+        }
+    };
+    
+    match env.new_string(rust_str) {
+        Ok(jstring) => {
+            unsafe { c2pa_string_free(c_str) };
+            jstring.into_raw()
+        }
+        Err(_) => {
+            unsafe { c2pa_string_free(c_str) };
+            ptr::null_mut()
+        }
+    }
+}
+
+/// Returns a version string for logging
+#[no_mangle]
+pub extern "system" fn Java_org_c2pa_C2PA_getVersion(
+    mut env: JNIEnv,
+    _class: JClass,
+) -> jstring {
+    let version = unsafe { c2pa_version() };
+    c_string_to_java_string(&mut env, version)
+}
+
+/// Returns the last error message
+#[no_mangle]
+pub extern "system" fn Java_org_c2pa_C2PA_getError(
+    mut env: JNIEnv,
+    _class: JClass,
+) -> jstring {
+    let error = unsafe { c2pa_error() };
+    c_string_to_java_string(&mut env, error)
+}
+
+/// Load Settings from a string
+#[no_mangle]
+pub extern "system" fn Java_org_c2pa_C2PA_loadSettings(
+    mut env: JNIEnv,
+    _class: JClass,
+    settings: JString,
+    format: JString,
+) -> jint {
+    let settings_cstr = java_string_to_cstring_jint!(env, settings);
+    let format_cstr = java_string_to_cstring_jint!(env, format);
+    
+    unsafe { c2pa_load_settings(settings_cstr.as_ptr(), format_cstr.as_ptr()) }
+}
+
+/// Returns a ManifestStore JSON string from a file path
+#[no_mangle]
+pub extern "system" fn Java_org_c2pa_C2PA_readFile(
```

Comment:
Ack - but we should consider using readers to do that, as they are an important piece of the API surface (and it reduces the exposed API surface too, so people use the API's main concepts).